### PR TITLE
Allow docker to run e2e tests

### PR DESCRIPTION
**Pull request by PierreRambaud**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Add a proper docker-compose.yml file to run tests in isolate situation.
| Type?         | improvement
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| How to test?  | Follow DOCKER.md file.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9400)

### CODE REVIEW

**Comment by PierreRambaud** [comment]
@matks @mickaelandrieu Needs tests and feedback for improvement.

**Comment by matks** [comment]
When the PR is not in WIP anymore, it needs to modify the file `tests/E2E/README.md`. The 2 files README.md and DOCKER.md are confusing right now as there is duplicated content.

**Comment by PierreRambaud** [comment]
@matks Done for the Readme ;)

File: tests/E2E/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      - "prestashop"
+    volumes:
+      - /dev/shm:/dev/shm
+    networks:
+      default:
+        aliases:
+          - selenium-chrome
+
+  db:
+    image: mysql:5.7.22
```
Comment (Quetzacoalt91) [inline tests/E2E/docker-compose.yml:54]: Are we sure this tag will always be available? Isn't `mysql:5.7` safer?

Comment (PierreRambaud) [reply tests/E2E/docker-compose.yml:54]: This tag contains a fix because of a bug with latest version of Mysql container. It's weird but recommended :/

Comment (Quetzacoalt91) [reply tests/E2E/docker-compose.yml:54]: Let's keeping it until the day we cannot use it anymore then. :)

**Comment by PierreRambaud** [comment]
I remove the WIP status, hope everyone will be able to run e2e tests from now. Don't hesitate to tell me if something is missing.

File: install-dev/classes/controllerConsole.php [hunk 2]
```diff
@@ -153,10 +153,13 @@ public function printErrors()
             if (!is_array($errors)) {
                 $errors = array($errors);
             }
-            echo 'Errors :'."\n";
+            echo 'Errors :'. PHP_EOL;
             foreach ($errors as $error_process) {
+                if (!is_array($error_process)) {
```
Comment (PierreRambaud) [inline install-dev/classes/controllerConsole.php:158]: @Quetzacoalt91 Having some errors here, don't know why $error_process isn't an array.

Comment (Quetzacoalt91) [reply install-dev/classes/controllerConsole.php:158]: I guess the array we merged does not always respect this rule.

File: tests/E2E/.docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+# Install dependencies
+RUN apt-get update
+RUN apt-get install -y apt-utils
+RUN apt install -y \
+    gnupg2 \
+    curl \
+    git \
+    software-properties-common \
+    nodejs
```
Comment (Quetzacoalt91) [inline tests/E2E/.docker/Dockerfile]: Nodejs is not needed before the new repository is declared.

File: tests/E2E/.docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+# Repository update
+WORKDIR /var/www/html/tests/E2E
+
+RUN usermod -u 1000 www-data
+
+COPY prepare-prestashop.sh /tmp/prepare-prestashop.sh
+
+COPY /prestashop/wait-for-it.sh /tmp/
+COPY run-tests.sh /tmp/
```
Comment (Quetzacoalt91) [inline tests/E2E/.docker/Dockerfile:31]: Hmm, I wonder if this build can't be improved. The reason could be avoiding to download too many images, bu you could directly start from a nodejs image and copy the e2e tests in it. :)

File: tests/E2E/.docker/Dockerfile (hunk 3 above)
Comment (PierreRambaud) [reply tests/E2E/.docker/Dockerfile]: Thanks :) Nice catch

File: tests/E2E/.docker/Dockerfile (hunk 4 above)
Comment (PierreRambaud) [reply tests/E2E/.docker/Dockerfile:31]: I tried with node image, but forget why I move to `prestashop/base:7.2-apache` ><

File: tests/E2E/.docker/prestashop/docker_run.sh [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      mysqladmin -h $DB_SERVER -P $DB_PORT -u $DB_USER -p$DB_PASSWD drop $DB_NAME --force
+      mysqladmin -h $DB_SERVER -P $DB_PORT -u $DB_USER -p$DB_PASSWD create $DB_NAME --force
+    fi
+  fi
+
+  if [ "$PS_DOMAIN" = "<to be defined>" ]; then
+    export PS_DOMAIN=$(hostname -i)
+  fi
+
+  pushd /var/www/html
```
Comment (Quetzacoalt91) [inline tests/E2E/.docker/prestashop/docker_run.sh]: `prestashop_1       | /tmp/docker_run.sh: 70: /tmp/docker_run.sh: pushd: not found`

File: tests/E2E/.docker/prestashop/docker_run.sh [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  echo "\n* Install Dependencies...";
+  /usr/bin/composer install
+  popd
```
Comment (Quetzacoalt91) [inline tests/E2E/.docker/prestashop/docker_run.sh]: `prestashop_1       | /tmp/docker_run.sh: 73: /tmp/docker_run.sh: popd: not found`

File: tests/E2E/.docker/prestashop/docker_run.sh (hunk 5 above)
Comment (PierreRambaud) [reply tests/E2E/.docker/prestashop/docker_run.sh]: Should be ok now, maybe shebang problem :)

File: tests/E2E/.docker/prestashop/docker_run.sh (hunk 6 above)
Comment (PierreRambaud) [reply tests/E2E/.docker/prestashop/docker_run.sh]: Should be ok now, maybe shebang problem :)

**Comment by Quetzacoalt91** [comment]
Still broken unfortunately

```
prestashop_1       | wait-for-it.sh: db:3306 is available after 16 seconds
prestashop_1       | \n* Remove PrestaShop configuraiton files...
prestashop_1       | /tmp/docker_run.sh: line 23: [: argument expected
prestashop_1       | \n* Installing PrestaShop, this may take a while ...
prestashop_1       | \n* Drop & recreate mysql database...
prestashop_1       | mysqladmin: DROP DATABASE prestashop failed;
prestashop_1       | error: 'Can't drop database 'prestashop'; database doesn't exist'
prestashop_1       | /var/www/html /var/www/html
prestashop_1       | \n* Install Dependencies...
prestashop_1       | Do not run Composer as root/super user! See https://getcomposer.org/root for details
prestashop_1       | Loading composer repositories with package information
prestashop_1       | Installing dependencies (including require-dev) from lock file
prestashop_1       | Package operations: 8 installs, 2 updates, 0 removals
prestashop_1       |   - Updating prestashop/ps_themecusto (v1.0.6 => v1.0.8): Downloading (100%)         
prestashop_1       |   - Installing symfony/polyfill-php72 (v1.8.0): Downloading (100%)         
prestashop_1       |   - Installing php-cs-fixer/diff (v1.3.0): Downloading (100%)         
prestashop_1       |   - Installing composer/xdebug-handler (1.1.0): Downloading (100%)         
prestashop_1       |   - Installing composer/semver (1.4.2): Downloading (100%)         
prestashop_1       |   - Updating friendsofphp/php-cs-fixer (v1.13.3 => v2.12.2): Downloading (100%)         
prestashop_1       |   - Installing analog/analog (1.0.10-stable): Downloading (100%)         
prestashop_1       |   - Installing greenlion/php-sql-parser (v4.1.2): Downloading (100%)         
prestashop_1       |   - Installing prestashop/gsitemap (v3.2.2): Downloading (100%)         
prestashop_1       | 
prestashop_1       | /var/www/html
prestashop_1       | \n* Install PrestaShop...
prestashop_1       |                                                                    
prestashop_1       |   [RuntimeException]                                               
prestashop_1       |   modules/ps_customtext/ does not exist and could not be created.  
prestashop_1       |                                                                    
prestashop_1       | 
prestashop_1       | install [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [--no-suggest] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--ignore-platform-reqs] [--] [<packages>]...
prestashop_1       | 
```

File: tests/E2E/DOCKER.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+Docker should start building (if it's the first time for these images) and running with the containers in the background.
+Wait some seconds
+
+## Run tests
+
+```bash
+# Go inside tests container
+$ docker-compose exec tests bash
+$ /tmp/run-tests.sh
```
Comment (Quetzacoalt91) [inline tests/E2E/DOCKER.md]: Could be 

```bash
$ docker-compose exec tests bash /tmp/run-tests.sh
```

Comment (PierreRambaud) [reply tests/E2E/DOCKER.md]: Done

**Review by Quetzacoalt91 — APPROVED** [review]

**Comment by Quetzacoalt91** [comment]
Tested locally, merging.

Thanks @PierreRambaud

**Comment by mbadrani** [comment]
@fatmaBouchekoua any problem in this pr?

**Comment by fatmaBouchekoua** [comment]
I have this error when I run this command: `docker-compose up`
`tests_1            | wait-for-it.sh: timeout occurred after waiting 360 seconds for prestashop-web:80
tests_1            | wait-for-it.sh: strict mode, refusing to execute subprocess
e2e_tests_1 exited with code 124`

**Pull request merged**